### PR TITLE
Removed Gtk2hs dependencies from System.Taffybar.Context

### PR DESCRIPTION
**Pull request by Tordek**
This removes, as far as I can see, all references to Gtk2hs in Context.

This does raise some deprecation warnings related to hBox.

### CODE REVIEW

**Comment by IvanMalison** [comment]
@Tordek , while I suppose that this WILL compile, doesn't this mean that none of the existing widgets will work at all? I suppose they just need toGIWidget to be applied.

Lets include the transformations as part of this change?

**Comment by Tordek** [comment]
This should be all of the Widgets; most of them are Gtk2hs-free by transitivity (i.e., they only use gtk2hs-free generics), a few just cheat and do "toGIWidget".

All of the widgets containing toGIWidget, then (theoretically, that would mean the only thing remaining before getting rid of Gtk2hs) are:

src/System/Taffybar/Widget/FreedesktopNotifications.hs
src/System/Taffybar/Widget/Systray.hs
src/System/Taffybar/Widget/Layout.hs
src/System/Taffybar/Widget/Decorators.hs
src/System/Taffybar/Widget/XDGMenu/MenuWidget.hs
src/System/Taffybar/Widget/Generic/PollingGraph.hs
src/System/Taffybar/Widget/Generic/VerticalBar.hs
src/System/Taffybar/Widget/Generic/ChannelGraph.hs
src/System/Taffybar/Widget/Generic/Icon.hs
src/System/Taffybar/Widget/SimpleClock.hs
src/System/Taffybar/Widget/Workspaces.hs

File: src/System/Taffybar/Context.hs [hunk 1]
```diff
@@ -166,24 +166,22 @@ buildBarWindow context barConfig = do
       printf "Building bar window with StrutConfig: %s" $
       show $ strutConfig barConfig
 
-  window <- Gtk.windowNew
+  window <- Gtk.windowNew Gtk.WindowTypeToplevel
   box <- Gtk.hBoxNew False $ fromIntegral $ widgetSpacing barConfig
-  _ <- widgetSetClass box "taffy-box"
+  _ <- widgetSetClassGI box (T.pack "taffy-box")
```
[the remark this answers is not in the document]
Comment (IvanMalison) [reply src/System/Taffybar/Context.hs]: Yeah, we should use overloaded strings. I'll do this myself though.

File: src/System/Taffybar/SimpleConfig.hs [hunk 2]
```diff
@@ -135,7 +136,10 @@ simpleTaffybar :: SimpleTaffyConfig -> IO ()
 simpleTaffybar conf = dyreTaffybar $ toTaffyConfig conf
 
 getMonitorCount :: IO Int
-getMonitorCount = screenGetDefault >>= maybe (return 0) screenGetNMonitors
+getMonitorCount = do
```
Comment (IvanMalison) [inline src/System/Taffybar/SimpleConfig.hs]: Why was this changed so much? Just to add the fromIntegral? Why not just add a fromIntegral <$> out front?

**Comment by IvanMalison** [comment]
@Tordek This is great. How much have you tested this?

> All of the widgets containing toGIWidget, then (theoretically, that would mean the only thing remaining before getting rid of Gtk2hs) are:

> src/System/Taffybar/Widget/FreedesktopNotifications.hs
> src/System/Taffybar/Widget/Systray.hs
> src/System/Taffybar/Widget/Layout.hs
> src/System/Taffybar/Widget/Decorators.hs
> src/System/Taffybar/Widget/XDGMenu/MenuWidget.hs
> src/System/Taffybar/Widget/Generic/PollingGraph.hs
> src/System/Taffybar/Widget/Generic/VerticalBar.hs
> src/System/Taffybar/Widget/Generic/ChannelGraph.hs
> src/System/Taffybar/Widget/Generic/Icon.hs
> src/System/Taffybar/Widget/SimpleClock.hs
> src/System/Taffybar/Widget/Workspaces.hs

Of these, the only that will present a significant amount of difficulty is the workspaces widget. There is a lot of code there, and I think that is honestly the largest hurdle.

**Comment by IvanMalison** [comment]
@Tordek I've made some changes at

https://github.com/taffybar/taffybar/tree/useGIGTKByDefault

I'm getting quite a few segfaults with these changes, but I suspect that this is a general issue with the gtk2hs gi-gtk interop that is simply made worse by these changes.

I think the best thing to do for now might be to use this new branch (useGIGTKByDefault) that I've created for intermediate steps until we have completely eliminated gtk2hs.

If you can offer your help, I think we can get this done in the next week or so.

I can probably tackle the workspaces module, which is going to be the thing which requires the most domain specific knowledge

**Comment by Tordek** [comment]
> How much have you tested this?

Not really anything beyond checking they compile; this was a quick thing, not a smart thing.

> I'm getting quite a few segfaults with these changes

I'm going to suspect this is caused by a few fromGIWidget calls here and there, there's a lot of back-and-forth and I'm guessing some pointers gets disowned and reused.

> If you can offer your help, I think we can get this done in the next week or so.

Sure!

> I can probably tackle the workspaces module, which is going to be the thing which requires the most domain specific knowledge

Awesome. I have a couple of questions... I think it was: How do I replace the `on` cases? with `onButtonPressed` and the like? and there's one, I don't remember where, where it goes `on hbox buttonPressed $ on [SingleClick]` or something along those lines?

**Comment by IvanMalison** [comment]
@Tordek 

Yeah the on calls needs to be replaced with

Gtk.onWidget...

You'll need to look at the documentation at 

http://hackage.haskell.org/package/gi-gtk for that

I just finished porting workspaces and the changes are up at the new branch, so you can take a look at that. Unfortunately, I'm still getting quite a few segfaults.

One candidate for this issue could be that we still intialize taffybar using gtk2hs:

https://github.com/travitch/taffybar/blob/f40d989593da0780c071b06456f2c088b198bc01/src/System/Taffybar.hs#L179

I'm going to work on porting that over now and see if it fixes anything.

**Comment by IvanMalison** [comment]
@Tordek 
Hopefully I'm not speaking to soon, but it seems like changing the initialization code may have done the trick. Been running for the last 15ish minutes and no crashes.
